Add Generic dataclasses

mvanderlee opened the pull request:
Building on the fantastic work of @onursatici and @dairiki in PRs #172 and #232

Currently broken on py3.6, but once #257 is merged I'll rebase and thus remove py3.6 and py3.7 support.


I've tested this with tox on python versions 38, 39, 310, 311, 312

Spoonrad commented:
Exciting PR, I've been waiting a long time for generic support in dataclasses :) Do you have any update please?

mvanderlee commented:
@dairiki @lovasoa This is ready for review.

@Spoonrad Consider this the update

mvanderlee commented:
I came up with 1 additional usecase, with 3 solutions that are somewhat related to Annotated and Generics. 
I have existing NewTypes like this:
```python
DelimitedListStr = NewType('DelimitedListStr', list[str], field=lambda *args, **kwargs: DelimitedList(mf.String(), *args, **kwargs))
```

I came up with these 3 solutions for this. One uses generics, the other doesn't.
1. Using Generics:
```python
DelimitedListStr = Annotated[List[str], DelimitedList[mf.String]]
```

2. Using partial:
```python
DelimitedListStr = Annotated[List[str], functools.partial(DelimitedList, mf.String)]
```
3. Using callable:
```python
DelimitedListStr = Annotated[List[str], lambda *args, **kwargs: DelimitedList(mf.String, *args, **kwargs)]
```

Support for each has been added, including unittests.
I also broke out the generic handling into it's own file.

dairiki commented:
> 2. Using partial:
> 
> ```python
> DelimitedListStr = Annotated[List[str], functools.partial(DelimitedList, mf.String)]
> ```
> 
> 3. Using callable:
> 
> ```python
> DelimitedListStr = Annotated[List[str], lambda *args, **kwargs: DelimitedList(mf.String, *args, **kwargs)]
> ```

As API choices, these make me nervous.

Recognizing generic callables as (potential) schema field annotations, and then testing them with a zero-arg call, as is done here

https://github.com/lovasoa/marshmallow_dataclass/blob/4531c35cf5f1b50e1dbc793395377a380aae4bf6/marshmallow_dataclass/__init__.py#L1090-L1097

seems both confusing and dangerous.
- The test call uses potentially different arguments than the callable is intended to be called with.
   -  The programmer must know to support zero-arg calls
   -  There is no guarantee that the real call (with arguments) will return the same thing as the zero-arg test call
- The callable (which could be an annotation unrelated to marshmallow_dataclass) might have side-effects

----

Can the same functionality be obtained by creating a custom Marshmallow Field subclass?  E.g. something like:

```py
class DelimitedStrListField(DelimitedList):
    def __init__(self, *args, **kwargs):
        super().__init__(mf.String, *args, **kwargs)

DelimitedListStr = Annotated[List[str], DelimitedStrListField]
```

mvanderlee commented:
Right.. We don't yet know if the annotation is supposed to be a marshmallow Field compatible callable. So we could be calling anything including `eraseDiskDrive()`. 

The class approach certainly works, I was trying to use a one liner. But the `partial` approach works for that, and is safer from an API perspective. 
We _could_ introduce a special callable wrapper, so the user can use a callable but has to explicitly tell us about it.
```python
DelimitedListStr = Annotated[List[str], MaFieldCallable(lambda *args, **kwargs: DelimitedList(mf.String, *args, **kwargs))]
```
but I don't see the value in that. over using either the `partial` or custom class approach.

dairiki commented:
> The class approach certainly works, I was trying to use a one liner. But the `partial` approach works for that, and is safer from an API perspective.

I hadn't noticed that you explicitly check for a partial applied to a Field ([here](https://github.com/lovasoa/marshmallow_dataclass/blob/4531c35cf5f1b50e1dbc793395377a380aae4bf6/marshmallow_dataclass/__init__.py#L759)).  That is safe.

I still think it adds unnecessary clutter to the API. 

Also, allowing a `partial` but not an equivalent arbitrary callable seems potentially confusing.
It doesn't add any capability to the API, other than to enable specific one-liners.

> We could introduce a special callable wrapper, so the user can use a callable but has to explicitly tell us about it.

A value of that would be that it is more explicitly readable: it makes the intent of the annotation more clear to the unindoctrinated reader.

----

I'm in favor of stopping at recognizing annotations that are either subclasses or instances of `Field`.  That's an understandable, simple-to-explain API that's fairly self-evident, even to those unfamiliar with it.

I don't really see the extra three lines required to define a custom Field subclass as too verbose.  (As well, doing so provides a good place to add a docstring, if that would help clarify the intent.)

mvanderlee commented:
I'll switch to using the generic approach myself. So I'm fine with removing the `partial` approach.
I mostly wanted to support an undocumented feature that works with NewType. 

>  allowing a partial but not an equivalent arbitrary callable seems potentially confusing.
Agreed.  I did investigate further but couldn't find a satisfactor way to get typehints of lambdas, even when using the Callable typehint.

While I agree that 3 lines isn't bad, given that I have multiple one lines currently, each would take up 3 + 2 blank lines (style guide) So instead of 5 lines, it'd be 25 lines. Which is harder to read at a glance. 

But again, generics fixes my usecase and I can't find a satisfactory answer to the valid concerns being brought up.

dairiki commented on marshmallow_dataclass/generic_resolver.py:
```diff
@@ -0,0 +1,193 @@
+    defined as `class A(Generic[T])`, this method will return true if `A[int]` is passed
+    """
+    is_generic = is_generic_type(clazz)
+    type_arguments = typing_extensions.get_args(clazz)
```
Can we just use `typing.get_args` here?

Then, just one-line this, with short-circuit:

```py
return is_generic_type(clazz) and len(get_args(clazz)) > 0
```

dairiki replied:
In general, can we use `typing.get_args` in place of `typing_extensions.get_args` (and `typing.get_origin` in place of typing_extensions.get_origin`) everywhere?

Both `get_origin` and `get_args` are in `typing` as of python 3.8.

mvanderlee replied:
They are both in 3.8, but when I use the `typing` functions the tests break. The implementation appears incomplete.
From 3.9 onwards the `typing` functions work fine.

dairiki replied:
Now I'm curious as to what the difference is.

What happens if one just accesses `.__origin__` and `.__args__` directly?  ([ref](https://docs.python.org/3/library/stdtypes.html#genericalias.__origin__))

mvanderlee replied:
I attributed the failed test to the wrong root cause.

Py 3.8 doesn't support Annotated, so we import it from `typing_extension` Therefore we need to use `get_origin` and `get_args` from typing_extension as well.

mvanderlee replied:
We could do this:
```python
if sys.version_info >= (3, 9):
    from typing import Annotated, get_args, get_origin
else:
    from typing_extensions import Annotated, get_args, get_origin
```

And just use the function without module in the code.

dairiki replied:
> Py 3.8 doesn't support Annotated, so we import it from `typing_extension` Therefore we need to use `get_origin` and `get_args` from typing_extension as well.

So it only matters when dealing with `Annotated`s?

mvanderlee replied:
Correct.

dairiki reviewed (commented):
I'm not sure that this review is complete, but I'm about to leave on a road-trip, so wanted to get this out.

mvanderlee commented on marshmallow_dataclass/__init__.py:
```diff
@@ -44,15 +44,9 @@ class User:
 from functools import lru_cache, partial
+from typing import Any, Callable, Dict, FrozenSet, Generic, List, Mapping
+from typing import NewType as typing_NewType
 from typing import (
```
Black did it automatically. I believe it's caused by the aliasing.

dairiki replied:
> Black did it automatically. I believe it's caused by the aliasing.

Black seems to be agnostic on this?

Playing with black 24.4.2, it seems to accept either the three-import version or the single-import version.  It seems to pass either of these without reformatting them:

```py
from typing import Any, Callable, Dict, FrozenSet, Generic, List, Mapping
from typing import NewType as typing_NewType
from typing import (
    Optional,
    Sequence,
    Set,
    Tuple,
    Type,
    TypeVar,
    Union,
    cast,
    get_type_hints,
    overload,
)
```
or

```py
from typing import (
    Any,
    Callable,
    Dict,
    FrozenSet,
    Generic,
    List,
    Mapping,
    NewType as typing_NewType,
    Optional,
    Sequence,
    Set,
    Tuple,
    Type,
    TypeVar,
    Union,
    cast,
    get_type_hints,
    overload,
)
```

Ref: https://github.com/psf/black/issues/333#issuecomment-1153637034

mvanderlee replied:
Well.. you're right. I didn't do it by hand :). But fixed now regardless.

mvanderlee commented on marshmallow_dataclass/generic_resolver.py:
```diff
@@ -0,0 +1,193 @@
+
+def is_generic_type(clazz: type) -> bool:
+    """
+    typing_inspect.is_generic_type explicitly ignores Union, Tuple, Callable, ClassVar
```
I'm not sure what change is being asked for here? A name change, or change in logic maybe?

I can't think of any usecase where we'd need to handle ClassVar and Callable to generate a marshmallow object, so I can add exclusion of them. 
But Tuple and Union are standard marshmallow fields, so that makes more sense to support.

dairiki replied:
I'm not sure what change is best here, either.  Clearly, the current function behavior does not match its name. (So, that is confusing.)

I'm not sure this function is necessary.  It seems most uses of it are on objects that are already known (or assumed to be dataclasses), in which case all these edge cases are moot.  (The other place it is used is on what are potentially generic marshmallow `Field`s.)

Note that unions are special-cased in `_field_for_union_type` and tuples (as well as all the "well known" containers — dicts, lists, etc) are special-cased in `_field_for_generic_type`.

Which brings up: `_field_for_generic_type` should probably be renamed to `_field_for_container_type` (or some such), since it doesn't really have anything to do with the support for generics added by this PR.

mvanderlee replied:
I've renamed the relevant functions and variables to `container`

I've dug a little deeper and we only need the `is_generic_type` to see if we need to resolve typevars. So I'll rename it to  `may_contain_typevars(clazz)` and only use it where applicable. And use `typing_inspect.is_generic_type` elsewhere

mvanderlee replied:
@dairiki I'm going to mark this thread as resolved. Re-open if you disagree.

mvanderlee commented on marshmallow_dataclass/__init__.py:
```diff
@@ -996,6 +1023,66 @@ def _get_field_default(field: dataclasses.Field):
+    obj,
+    schema_ctx: _SchemaContext,
+) -> type:
+    """typing.get_type_hints doesn't work with generic aliasses. But this 'hack' works."""
```
Hmm, the forward refs have already been resolved. This really exists purely to get the typehint for generics.
maybe `_get_type_hint_of_generic_object` ?


```python
import typing

T = typing.TypeVar('T')

class A(Generic[T]):
  a: T

class B(A[int]):
  pass

print(typing.get_type_hints(B))
print(typing.get_type_hints(A[int]))

>=========================
{'a': ~T}
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[51], [line 12](vscode-notebook-cell:?execution_count=51&line=12)
      [9](vscode-notebook-cell:?execution_count=51&line=9)   pass
     [11](vscode-notebook-cell:?execution_count=51&line=11) print(typing.get_type_hints(B))
---> [12](vscode-notebook-cell:?execution_count=51&line=12) print(typing.get_type_hints(A[int]))

File ~\.pyenv\pyenv-win\versions\3.11.3\Lib\typing.py:2347, in get_type_hints(obj, globalns, localns, include_extras)
   ~/.pyenv/pyenv-win/versions/3.11.3/Lib/typing.py:2345)         return {}
   ~/.pyenv/pyenv-win/versions/3.11.3/Lib/typing.py:2346)     else:
-> ~/.pyenv/pyenv-win/versions/3.11.3/Lib/typing.py:2347)         raise TypeError('{!r} is not a module, class, method, '
   ~/.pyenv/pyenv-win/versions/3.11.3/Lib/typing.py:2348)                         'or function.'.format(obj))
   ~/.pyenv/pyenv-win/versions/3.11.3/Lib/typing.py:2349) hints = dict(hints)
   ~/.pyenv/pyenv-win/versions/3.11.3/Lib/typing.py:2350) for name, value in hints.items():

TypeError: __main__.A[int] is not a module, class, method, or function.
```

```python
def _get_generic_type_hints(obj) -> type:
    """typing.get_type_hints doesn't work with generic aliases. But this 'hack' works."""

    class X:
        x: obj  # type: ignore[name-defined]

    return typing.get_type_hints(X)['x']

print(_get_generic_type_hints(B))
print(_get_generic_type_hints(A[int]))

>=========================
<class '__main__.B'>
__main__.A[int]
```

dairiki replied:
In your last examples, `_get_generic_type_hints` is just the identity, right?

```py
>>> _get_generic_type_hints(B) is B
True
>>> _get_generic_type_hints(A[int]) is A[int]
True
```

But, the value of `_get_generic_type_hints` comes in resolving forward references:
```
>>> _get_generic_type_hints(A["int"])
__main__.A[int]
>>> _get_generic_type_hints(A["int"]) is A[int]
False
```

dairiki replied:
Maybe what is really wanted is (untested)
```py
def _get_generic_type_hints(generic_alias):
    class X(generic_alias): pass

    return typing.get_type_hints(X)
```

Or, in one line:
```
    return typing.get_type_hints(types.new_class("_", bases=(generic_alias,)))
```

mvanderlee replied:
Wow, good catch! You're absolutely right.

1. This just worked 🤦‍♂️ I clearly spend too long trying to make it work to see the obvious solution.
```python
(
    type_hints[field.name]
    if not is_generic_type(clazz)
    else field.type
)
```

2. But didn't work for `A["int"]` and no tests caught that yet. So I'll add a tests and rename the function as per your recommendation.

dairiki replied:
And, if that works, maybe just roll this into `_get_type_hints` so that there's a single `_get_type_hints` function that can be used for any dataclass, be it generic or not.

mvanderlee replied:
> ```python
> def _get_generic_type_hints(generic_alias):
>     class X(generic_alias): pass
>
>     return typing.get_type_hints(X)
> ```

Does not work.
```python
print(_get_generic_type_hints(A[int]))

>=====================
{'a': ~T}
```

Instead of `{'a': __main__.A[int]}`

dairiki replied:
> Does not work.

That's too bad.

I still think that refactoring so that `_get_type_hints` will work with generic aliases of dataclasses as well as plain dataclasses is probably worth it.  It will move all the `is_generic_type` special-casing out of 

https://github.com/lovasoa/marshmallow_dataclass/blob/7ac088d6086c5e4e2313bced99f1454204834563/marshmallow_dataclass/__init__.py#L583-L603

mvanderlee replied:
I don't disagree but I've already sunk too many hours into this particular problem. If it is even possible, then I don't have the knowledge to take this further.

mvanderlee replied:
@dairiki I found a way around it by resolving the forward refs when we get the dataclass fields. We now no longer call `get_type_hints` at all anymore as it's relevant code is internalized. 

We were already looping over the mro and fields just like `get_type_hints`.

mvanderlee commented:
Bump.

Syntaf commented:
+1 to this PR, love this @mvanderlee -- we were just recently dealing with this issue and ended up creating our own bespoke field that handled this, but if this gets merged in we'd love to bump our version and use this.

lovasoa commented:
@mvanderlee I'm really sorry, I don't have a lot of bandwidth. 

@dairiki , maybe you can review this?

dairiki commented:
> @mvanderlee I'm really sorry, I don't have a lot of bandwidth.
> 
> @dairiki , maybe you can review this?

I apologize as well. For the past month I've been mostly out of town and swamped with other concerns and activities.
I'm home again now, but still trying to catch up.  I will try to find time for this in the next couple of weeks (but no guarantees).

dairiki commented on marshmallow_dataclass/__init__.py:
```diff
@@ -214,12 +229,18 @@ def dataclass(
     )
 
     def decorator(cls: Type[_U], stacklevel: int = 1) -> Type[_U]:
+        if cls is not None:
```
Why the `if cls is not None` check?

Shouldn't `decorator` fail if called on `None`?
(I believe `_check_decorated_type(None)` will (correctly?) raise `TypeError("expected a class not None")`.)

mvanderlee replied:
I don't recall tbh. Removing it didn't break any tests, and I've been careful to write tests everytime I ran into an issue. 
And I would agree that the TypeError is as expected.

Removed if statement.

dairiki commented on marshmallow_dataclass/__init__.py:
```diff
@@ -214,12 +229,18 @@ def dataclass(
     if _cls is None:
         return decorator
+
+    if _cls is not None:
+        _check_decorated_type(_cls)
```
Is this call to `_check_decorated_type` necessary here?
The next line calls `decorator` which also will invoke `_check_decorated_type`, right?

(And if this isn't necessary, `_check_decorated_type` can be inlined into the body of `decorator`.)

mvanderlee replied:
Correct, removed

dairiki commented on marshmallow_dataclass/__init__.py:
```diff
@@ -268,6 +289,8 @@ def add_schema(_cls=None, base_schema=None, cls_frame=None, stacklevel=1):
     """
 
     def decorator(clazz: Type[_U], stacklevel: int = stacklevel) -> Type[_U]:
+        _check_decorated_type(clazz)
```
`_check_decorate_type` requires only that `isinstance(clazz, type)`.
Do we want to require that clazz is a dataclass (`isinstance(clazz, type) and dataclasses.is_dataclass(clazz)`) here?

mvanderlee replied:
I don't think so because we already have a big warning when a class is not a dataclass in `_internal_class_schema`
So non-dataclasses are allowed, but not supported

dairiki replied:
I've just run a simple test.

You are correct in that decorating non-data class classes with `add_schema` works — at least in simple cases.  (That doesn't necessarily mean that we should allow it.)

It appears, however, that, as things stand in this PR, no big warning is emitted in that case.   Further investigation reveals that `get_resolved_dataclass_fields` "just works" (with no warnings emitted) even if its argument is not a dataclass.  (I have some suspicion that perhaps fields from the classes `__mro__` are not properly handled in that case, but I haven't looked close enough to say for sure.)

In any case, I think that either:
- We should disallow using the `add_schema` decorator on non-dataclasses.  (Why allow it if it's unsupported/untested?)
- Or, we need a test to ensure the warning is, in fact, emitted when it is.

mvanderlee replied:
I've had to do some digging.

1. In today's version, 8.7.1, I can call `class_schema(NonDataclass)` just fine. It only shows the warning if one of it's fields is not a dataclass. i.e.: Nested non dataclasses.
2. This goes back to when the warning was originally added: https://github.com/lovasoa/marshmallow_dataclass/commit/e31faa81a41691e8493ac18fb8f8aa44aeb52dbd
3. The behaviour still works the same with this PR.

I don't disagree with removing support for non-dataclasses, but don't see why that should be part of this PR.

mvanderlee commented:
Gentle bump

lovasoa commented:
@dairiki , feel free to merge if you think it's okay

dairiki commented:
@lovasoa @mvanderlee I apologize profusely for my lack of attention to this. (In my defense, I have been busy: travel, family and pet medical issues, and a lost month due to COVID infection.)

I've attempted to start a review on this a couple of times.  Each time I get bogged down by what seems to be mostly a re-implementation of `typing.get_type_hints` into the new code in [generic_resolver.py](https://github.com/lovasoa/marshmallow_dataclass/blob/d09312bb0430bacb42687fc545a1432a5aacc5b8/marshmallow_dataclass/generic_resolver.py).  (The re-implementation of `get_type_hints`, I think, is in [_resolve_typevars](https://github.com/lovasoa/marshmallow_dataclass/blob/d09312bb0430bacb42687fc545a1432a5aacc5b8/marshmallow_dataclass/generic_resolver.py#L136-L192).  I haven't had to time to fully digest all of what's in `generic_resolver.py`, so could be wrong.)

I haven't found a great way to phrase this, so I will just say it:
My main concern is that of maintenance headache.  I fear that the code in `generic_resolver.py` will be fragile with respect to upstream changes in implementation details of `typing`.  (Where our current use of `typing.get_type_hints`, as well as the `typing-inspect`, and `typeguard` packages mostly — or, at least, partially — isolate us from those changes.)

I could be wrong about that.  (I haven't yet fully digested the new code in `generic_resolver.py`.)

Or perhaps, the rest of you are not as bothered as I am with attempting to keep a level of abstraction (that is not under our maintenance) between our code and the implementation details of `typing`.

Feel free to discuss...  But that is what I've been getting stuck on.

dairiki commented:
> I fear that the code in `generic_resolver.py` will be fragile with respect to upstream changes in implementation details of `typing`.

As an example, how well does this PR cope with [PEP 696] (coming October 7 in Python 3.13)?
(I don't know the answer. It may work fine.)

In any case, since the 3.13 release is imminent, we should be testing with that, too.

[PEP 696]: https://peps.python.org/pep-0696/

mvanderlee commented:
The generic_resolver itself does a lot more than `get_type_hints` Without it only shallow generics can be supported and any nesting or overrides can't be resolved.

The merging with `get_type_hints` was an optimization to remove some looping over the fields, but can be easily reverted. https://github.com/mvanderlee/marshmallow_dataclass/commit/4ef0bdfa2eeea84cbda163371c2b12d577508428#diff-b59699b116d22f6ae7d49d5fc2084e62a695db0849c341c1a0840b7ce2202eefR139

The above 'merge' does break in 3.13 due to a warning being raised. The added type_params from 3.12 that will be required as an arg in `typing._eval_type` in 3.15 and trigger a warning in 3.13. It worked fine after I reverted to before that commit.

Adding support for TypeVars with defaults would probably be a fair bit of work mainly because of the recursion caused by allowing TypeVars as defaults. Or if the there are multiple defaults that should override each other.

mvanderlee commented:
@dairiki I have a second branch with the `get_type_hints` merge commit reverted and py 3.13 TypeVar defaults added:
https://github.com/mvanderlee/marshmallow_dataclass/tree/generic-dataclasses-get-type-hints

If that matches your expectations I can update this PR with that.

mvanderlee commented:
I've added Python 3.13 and reverted the `get_type_hints` commit. I'd love it if this could be wrapped up before end of year and will prioritize any further change requests.

mvanderlee commented:
@dairiki When do you think you'd have time to review?

mvanderlee commented:
@Spoonrad @Syntaf I'm tired of waiting and released a fork.  You may be interested in using it as well: https://pypi.org/project/marshmallow-dataclass2/
